File: leet/0098-validate-binary-search-tree/validate-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
    
    if root == nil {return true}
    
    if root.Left != nil {
        left := make([]*TreeNode, 1)
        left[0] = root.Left
        
        for i := 0; i < len(left); i++ {
            if left[i].Val >= root.Val {return false}
            
            if left[i].Left != nil {left = append(left, left[i].Left)}
            if left[i].Right != nil {left = append(left, left[i].Right)}
        }
    }
    
    if root.Right != nil {
        right := make([]*TreeNode, 1)
        right[0] = root.Right
        
        for i := 0; i < len(right); i++ {
            if right[i].Val <= root.Val {return false}
            
            if right[i].Left != nil {right = append(right, right[i].Left)}
            if right[i].Right != nil {right = append(right, right[i].Right)}
        }
    }
    
    return isValidBST(root.Left) && isValidBST(root.Right)
    
}

File: leet/0098-validate-binary-search-tree/validate-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    if (!root) return true
    
    let less = function (node, val){
        return !node || (node.val < val && less(node.left, val) && less(node.right, val))
    }
    
    let greater = function(node, val){
        return !node || (node.val > val && greater(node.left, val) && greater(node.right, val))
    }
    
    return less(root.left, root.val) && greater(root.right, root.val) && isValidBST(root.left) && isValidBST(root.right)
};
